xapian-applications/omega/clickmodel/simplifieddbn.cc

```diff
@@ -43,60 +43,61 @@ SimplifiedDBN::name()
 }
 
 static vector<string>
-get_docid_list(const vector<vector<string>> &sessions)
+get_docid_list(const vector<string> &session)
 {
     vector<string> docids;
     string docid;
-    for (size_t i = 0; i < sessions.size(); ++i) {
-	for (size_t j = 0; j <= sessions[i][DOCIDS].length(); ++j) {
-	    if (sessions[i][DOCIDS][j] != ',' && sessions[i][1][j] != '\0') {
-		docid += sessions[i][DOCIDS][j];
-	    } else {
-		docids.push_back(docid);
-		docid.clear();
-	    }
+    for (size_t j = 0; j <= session[DOCIDS].length(); ++j) {
+	char ch = session[DOCIDS][j];
+	if (ch != ',' && ch != '\0') {
+	    docid += ch;
+	} else {
+	    docids.push_back(docid);
+	    docid.clear();
 	}
     }
     return docids;
 }
 
 static vector<int>
-get_click_list(const vector<vector<string>> &sessions)
+get_click_list(const vector<string> &session)
 {
     vector<int> clicks;
     string clickstring, clickcount;
     int click = 0;
-    for (size_t i = 0; i < sessions.size(); ++i) {
-	for (size_t j = 0; j <= sessions[i][CLICKS].length(); ++j) {
-	    if (sessions[i][CLICKS][j] == '\0') {
-		clicks.push_back(click);
-	    } else if (sessions[i][CLICKS][j] != ',') {
-		// Get clickstring of the form "docid:click_count".
-		clickstring += sessions[i][CLICKS][j];
-
-		// Get click count string.
-		int delimiter_pos = clickstring.find(':');
-		clickcount = clickstring.substr(delimiter_pos + 1,
-			     clickstring.length() - 1 - delimiter_pos);
-
-		// Convert click count string to an integer.
-		stringstream ss(clickcount);
-		ss >> click;
-	    } else {
-		clickstring.clear();
-		clicks.push_back(click);
-	    }
+
+    for (size_t j = 0; j <= session[CLICKS].length(); ++j) {
+	char ch = session[CLICKS][j];
+	if (ch != ',' && ch != '\0') {
+	    // Get clickstring of the form "docid:click_count".
+	    clickstring += session[CLICKS][j];
+
+	    // Get click count string.
+	    size_t delimiter_pos = clickstring.find(':');
+	    clickcount = clickstring.substr(delimiter_pos + 1);
+
+	    // Convert click count string to an integer.
+	    stringstream ss(clickcount);
+	    ss >> click;
+	} else {
+	    clicks.push_back(click);
+	    clickstring.clear();
 	}
     }
     return clicks;
 }
 
 vector<vector<string>>
-SimplifiedDBN::build_sessions(const string logfile)
+SimplifiedDBN::build_sessions(const string &logfile)
 {
     ifstream file;
     file.open(logfile, ios::in);
 
+    if (!file) {
+	cerr << "ERROR: Specified file " << logfile;
+	throw runtime_error(" does not exist.");
```

Comment:
Surely this doesn't make sense — half the error message goes to `cerr`, but half is in the exception message. Probably the whole thing should be in the exception message, so that client code can catch it and do something appropriate.